src/client/java/teammates/client/scripts/DataMigrationForRubricQuestionsWithWeightsAttached.java

```diff
@@ -0,0 +1,112 @@
+package teammates.client.scripts;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.List;
+
+import com.googlecode.objectify.Key;
+
+import teammates.common.datatransfer.attributes.FeedbackQuestionAttributes;
+import teammates.common.datatransfer.questions.FeedbackQuestionType;
+import teammates.common.datatransfer.questions.FeedbackRubricQuestionDetails;
+import teammates.storage.entity.FeedbackQuestion;
+
+/**
+ * Script to load Rubric questions, and if the question has weights attached, converts the one-dimensional
+ * rubric weights list to a two-dimensional weight list.
+ */
+public class DataMigrationForRubricQuestionsWithWeightsAttached extends DataMigrationBaseScript<Key<FeedbackQuestion>> {
+
+    public static void main(String[] args) throws IOException {
+        new DataMigrationForRubricQuestionsWithWeightsAttached().doOperationRemotely();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected boolean isPreview() {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected List<Key<FeedbackQuestion>> getEntities() {
+        return ofy().load().type(FeedbackQuestion.class).filter("questionType", FeedbackQuestionType.RUBRIC).keys().list();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected boolean isMigrationNeeded(Key<FeedbackQuestion> questionKey) {
+        FeedbackQuestion question = ofy().load().key(questionKey).now();
+        FeedbackQuestionAttributes attr = FeedbackQuestionAttributes.valueOf(question);
+        FeedbackRubricQuestionDetails fqd = (FeedbackRubricQuestionDetails) attr.getQuestionDetails();
+
+        try {
+            Field hasAssignedWeights = fqd.getClass().getDeclaredField("hasAssignedWeights");
+            hasAssignedWeights.setAccessible(true);
+            if (!hasAssignedWeights.getBoolean(fqd)) {
+                return false;
+            }
+
+            Field rubricWeightsForEachCell = fqd.getClass().getDeclaredField("rubricWeightsForEachCell");
+            rubricWeightsForEachCell.setAccessible(true);
+            List<List<Double>> newWeights = (List<List<Double>>) rubricWeightsForEachCell.get(fqd);
+
+            // Question contains legacy data if rubricWeightsForEachCell list is empty.
+            return newWeights.isEmpty();
+        } catch (ReflectiveOperationException e) {
+            System.out.println(e.getMessage());
+            return true;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void printPreviewInformation(Key<FeedbackQuestion> questionKey) {
+        // nothing to do.
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void migrate(Key<FeedbackQuestion> questionKey) {
+        ofy().transact(() -> {
+            FeedbackQuestion question = ofy().load().key(questionKey).now();
+            question = updateWeights(question);
```

Comment:
As `isMigrationNeeded` and `migrate` are not wrapped in a transaction. An entity that is supposed to migrated may turn to be an entity that does not need migration, which means in our currently implementation, we may override the `rubricWeightsForEachCell` with old data.

I think the chance is kind of low (maybe?).

Not your fault @sukanta-27 but I think all the data migration script extend `DataMigrationBaseScript` get the same problem

Reply:
⬆️ This @sukanta-27

Reply:
>An entity that is supposed to migrated may turn to be an entity that does not need migration, which means in our currently implementation, we may override the rubricWeightsForEachCell with old data.

But if we look at the below implementation in `DataMigrationBaseScript` if `isMigrationNeeded` is set as `false` (L80-81), we continue to the next entity, the migration takes place at L89, then how can we override the values in `rubricWeightsForEachCell` list, as if the list is not empty, then `isMigrationNeeded` will not be `true`. 

https://github.com/TEAMMATES/teammates/blob/fd867cf408f64a9f94dbf2d4436527430f1361f1/src/client/java/teammates/client/scripts/DataMigrationBaseScript.java#L77-L94

Reply:
What if `rubricWeightsForEachCell` get modified (let's say become non-empty and we are not supposed to migrate it) when we are in L83 (actually any point between L80-88)?

Reply:
You mean if a user adds a new rubric question with non-empty `rubricWeightsForEachCell ` list while we are doing the migration?

Reply:
An existing rubric question's weight get modified.

Reply:
Hmmm. I understood the problem you are specifying. And I am not sure how we can handle that tbh :sweat_smile: I guess this means we will have to modify the base script as well as the script in this PR as you suggested. 

@xpdavid @damithc What should we do? Also maybe worth noting that from the previous script that we ran for finding how many negative weights are there, I guess there were a little over 1000 rubric questions with weights attached.

Reply:
If the chances are very low, we can tolerate it.

What we need w.r.t. abort/resume is something like this:
Say we need to process 200k entities which takes about 2 hrs but app engine disconnects after 1 hour (i.e., after processing 100k entities). We should be able to run the script again and process the remaining 100k in one more more hour i.e., second run should take 1 hour, not 2 hours. If the script takes one hour to go through the first 100k entities again, we will never be able to finish the migration as everytime we'll get disconnected after processing the first 100k entities.

Reply:
> @xpdavid @damithc What should we do? Also maybe worth noting that from the previous script that we ran for finding how many negative weights are there, I guess there were a little over 1000 rubric questions with weights attached.

Should be fine. No need to cater for a VERY unlikely data loss.